hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/SimpleConcurrentFileWritesConflictResolutionStrategy.java

```diff
@@ -0,0 +1,105 @@
+
+  @Override
+  public Stream<HoodieInstant> getCandidateInstants(HoodieActiveTimeline activeTimeline, HoodieInstant currentInstant,
+                                                 Option<HoodieInstant> lastSuccessfulInstant) {
+
+    // To find which instants are conflicting, we apply the following logic
+    // 1. Get completed instants timeline only for commits that have happened since the last successful write.
+    // 2. Get any scheduled or completed compaction or clustering operations that have started and/or finished
+    // after the current instant. We need to check for write conflicts since they may have mutated the same files
+    // that are being newly created by the current write.
+    // NOTE that any commits from table services such as compaction, clustering or cleaning since the
+    // overlapping of files is handled using MVCC.
+    Stream<HoodieInstant> completedCommitsInstantStream = activeTimeline
+        .getCommitsTimeline()
+        .filterCompletedInstants()
+        .findInstantsAfter(lastSuccessfulInstant.isPresent() ? lastSuccessfulInstant.get().getTimestamp() : HoodieTimeline.INIT_INSTANT_TS)
+        .getInstants();
+
+    Stream<HoodieInstant> compactionAndClusteringTimeline = activeTimeline
+        .getTimelineOfActions(CollectionUtils.createSet(REPLACE_COMMIT_ACTION, COMPACTION_ACTION))
+        .findInstantsAfter(currentInstant.getTimestamp())
+        .getInstants();
+    return Stream.concat(completedCommitsInstantStream, compactionAndClusteringTimeline);
+  }
+
+  @Override
+  public boolean hasConflict(HoodieCommitOperation thisOperation, HoodieCommitOperation otherOperation) {
+    // TODO : UUID's can clash even for insert/insert, handle that case.
+    Set<String> fileIdsSetForFirstInstant = thisOperation.getMutatedFileIds();
+    Set<String> fileIdsSetForSecondInstant = otherOperation.getMutatedFileIds();
+    Set<String> intersection = new HashSet<>(fileIdsSetForFirstInstant);
+    intersection.retainAll(fileIdsSetForSecondInstant);
+    if (!intersection.isEmpty()) {
+      LOG.warn("Found conflicting writes between first operation = " + thisOperation
+          + ", second operation = " + otherOperation + " , intersecting file ids " + intersection);
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public Option<HoodieCommitMetadata> resolveConflict(HoodieTable table,
+      HoodieCommitOperation thisOperation, HoodieCommitOperation otherOperation) {
+    // Since compaction is eventually written as commit, we need to ensure
+    // we handle this during conflict resolution and not treat the commit from compaction operation
+    // as a regular commit. Regular commit, deltacommits and replace are candidates for conflict
+    // replace is used for a) clustering without update support b) insert_overwrite both of which are
+    // candidates for conflict. We need to add CLUSTER here once it supports concurrent updates
```

Comment:
lets file a JIRA for clustering/concurrent updates.

Reply:
There is already one -> https://issues.apache.org/jira/browse/HUDI-1042